core/lib/rom/associations/definitions/abstract.rb

```diff
@@ -51,15 +56,45 @@ class Abstract
         #   @return [TrueClass,FalseClass] Whether custom view should override default one or not
         option :override, optional: true, default: -> { false }
 
-        alias_method :name, :as
-
+        # Instantiate a new association definition
+        #
+        # @param [Symbol] source The name of the source dataset
+        # @param [Symbol] target The name of the target dataset
+        # @param [Hash] options The option hash
+        # @option options [Symbol] :as The name of the association (defaults to target)
+        # @option options [Symbol] :relation The name of the target relation (defaults to target)
+        # @option options [Symbol] :foreign_key The name of a custom foreign key
+        # @option options [Symbol] :view The name of a custom relation view on the target's relation side
+        # @option options [TrueClass,FalseClass] :override Whether provided :view should override association's default view
+        #
         # @api public
-        def self.new(source, target, options = EMPTY_HASH)
-          super(
-            Name[source],
-            Name[options[:relation] || target, target, options[:as] || target],
-            options
-          )
+        def self.new(source, target, opts = EMPTY_HASH)
+          source_name = Relation::Name[source]
+          target_name = resolve_target_name(target, opts)
+          options = process_options(target_name, Hash[opts])
+
+          super(source_name, target_name, options)
+        end
+
+        # @api private
+        def self.resolve_target_name(target, options)
+          dataset = target
+          relation = options.fetch(:relation, target)
+
+          Relation::Name[relation, dataset, options[:as]]
+        end
+
+        # @api private
+        def self.process_options(target, options)
+          through = options[:through]
+
+          if through
+            options.update(through: ThroughIdentifier[through, target.relation, options[:assoc]])
+          end
+
+          options.update(name: target.relation)
+
+          options
```

Comment:
technically, you don't need this line, `updates` returns the updated hash anyway, otherwise, I'd prefer to go with `options[:name] = target.relation` because this won't allocate an intermediate hash instance '{ name: target.relation }'